### PR TITLE
Remove truncated backpropagation from loops

### LINKED ISSUE #6539: Should truncated_bptt_steps take effect during validation phase?
## 🐛 Bug

It appears that the Trainer flag `truncated_bptt_steps` doesn’t affect the validation phase.  Should it?  The problem I’m running into is that I need truncated_bptt_steps to virtually increase the length of the sequence I can fit into my GPU memory, but this purpose is defeated when the validation step doesn’t also make use of truncated_bptt_steps — in this case my memory is limited by my validation step, which attempts to process the full sequence in-memory.  

Is there a reason why validation step shouldn’t also make use of the truncated_bptt_steps flag?  Or is this just not finished yet?  I could probably manage to adapt the evaluation_loop code to copy the training_loop code regarding truncated_bptt_steps.  Would this be an acceptable addition?

See also: https://github.com/PyTorchLightning/pytorch-lightning/discussions/6483

## Please reproduce using the BoringModel

https://colab.research.google.com/drive/1JYAVdMy-UPg0Rzwpo3T9A-FmwrOdi_aR?usp=sharing

### Expected behavior

I expected the validation loop to also split the batch along the time dimension according to truncated_bptt_steps.

### LINKED ISSUE #15057: tbptt doesn't work with validation
### Bug description

The `tbptt_split_batch` function doesn't seem to be called before `validation_step` and `hiddens` also doesn't seem to be passed to the `validation_step` method of the lightning module.
Hence validating a model with a validation_step doesn't seems possible even if training can be achived using tbptt for timeseries data that's to big to fit in on the GPU/ where the loss calculation is to big to fit on the GPU. 

My current workarround is as follows:
```Python
def validation_step(self, batch, batch_idx):
        batch = self.tbptt_split_batch(batch, self.truncated_bptt_steps)
        hiddens = None
        losses = []

        for chung in batch:
            x = chung[0]  # x: [batch, seq, features]
            y_hat, hiddens = self(x, hiddens)  # y_hat: [batch, seq, features]
            loss = F.l1_loss(y_hat, x, reduction='sum')  # autoencoder model -> y_hat should be equal to x
            losses.append(loss)

        loss = sum([l/len(losses) for l in losses])
        self.log("val_loss", loss)
        return {"loss": loss}
```

However it would be great to have the same behaviour as with the `training_step` (if  `truncated_bptt_steps` is defined the chuncks are passed to `validation_step` as well as the hiddens from the last chunk).
I'm also not quite sure if my loss aggregation is the same as the one implemented for the training_step losses of the tbptt chunks (mean).

### How to reproduce the bug

```python
1. Use a timeseries dataset where a single batch/sample doesn't fit in Mem or can't perform backward because the sequence length is too big.
2. Define a model with `truncated_bptt_steps` set.
3. Call `trainer.fit(train_dl, val_dl)`
4. See Out of Memory exception because `tbptt_split_batch` isn't called on:
  a. the `sanity_check`
  b. the `validation_step`
5. remove `validation_step` from the model and the model model runs without problems because `tbptt_split_batch` is called in the train_loop.
```

### Error messages and logs

```
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
Running in `fast_dev_run` mode: will run the requested loop using 1 batch(es). Logging and checkpointing is suppressed.
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
  | Name    | Type | Params
---------------------------------
0 | encoder | RNN  | 3.4 K 
1 | decoder | RNN  | 324   
---------------------------------
3.7 K     Trainable params
0         Non-trainable params
3.7 K     Total params
0.015     Total estimated model params size (MB)
/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py:1892: PossibleUserWarning: The number of training batches (1) is smaller than the logging interval Trainer(log_every_n_steps=50). Set a lower value for log_every_n_steps if you want to see logs for the training epoch.
  rank_zero_warn(
Epoch 0:  50%|██████████████████████████████████████████████████████████                                                          | 1/2 [00:47<00:47, 47.35s/it, loss=863, split_idx=33, v_num=, train_step_loss (l1)=356.0Traceback (most recent call last):                                                                                                                                                                     | 0/1 [00:00<?, ?it/s]
  File "/home/username/cwd/scripts/./train.py", line 94, in <module>
    main(args)
  File "/home/username/cwd/scripts/./train.py", line 71, in main
    train(trainer_args=args, model=model, train_dl=train_dataloader, val_dl=validation_dataloader)
  File "/home/username/cwd/scripts/./train.py", line 56, in train
    trainer.fit(model, train_dataloaders=train_dl, val_dataloaders=val_dl)
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 696, in fit
    self._call_and_handle_interrupt(
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 650, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 735, in _fit_impl
    results = self._run(model, ckpt_path=self.ckpt_path)
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1166, in _run
    results = self._run_stage()
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1252, in _run_stage
    return self._run_train()
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1283, in _run_train
    self.fit_loop.run()
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/loops/loop.py", line 200, in run
    self.advance(*args, **kwargs)
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/loops/fit_loop.py", line 271, in advance
    self._outputs = self.epoch_loop.run(self._data_fetcher)
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/loops/loop.py", line 201, in run
    self.on_advance_end()
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/loops/epoch/training_epoch_loop.py", line 241, in on_advance_end
    self._run_validation()
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/loops/epoch/training_epoch_loop.py", line 299, in _run_validation
    self.val_loop.run()
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/loops/loop.py", line 200, in run
    self.advance(*args, **kwargs)
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/loops/dataloader/evaluation_loop.py", line 155, in advance
    dl_outputs = self.epoch_loop.run(self._data_fetcher, dl_max_batches, kwargs)
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/loops/loop.py", line 200, in run
    self.advance(*args, **kwargs)
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/loops/epoch/evaluation_epoch_loop.py", line 143, in advance
    output = self._evaluation_step(**kwargs)
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/loops/epoch/evaluation_epoch_loop.py", line 240, in _evaluation_step
    output = self.trainer._call_strategy_hook(hook_name, *kwargs.values())
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1704, in _call_strategy_hook
    output = fn(*args, **kwargs)
  File "/home/username/venv/lib/python3.9/site-packages/pytorch_lightning/strategies/strategy.py", line 370, in validation_step
    return self.model.validation_step(*args, **kwargs)
  File "/home/username/cwd/scripts/utils/deeplearning/models/autoencoder.py", line 53, in validation_step
    y_hat, hiddens = self(x, hiddens)  # y_hat: [batch, seq, features]
  File "/home/username/venv/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/username/cwd/scripts/utils/deeplearning/models/autoencoder.py", line 27, in forward
    _, enc_hiddens = self.encoder(x, enc_hiddens)  # hidden: [layer, batch, hidden]
  File "/home/username/venv/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/username/venv/lib/python3.9/site-packages/torch/nn/modules/rnn.py", line 471, in forward
    result = _VF.rnn_tanh(input, hx, self._flat_weights, self.bias, self.num_layers,
RuntimeError: CUDA out of memory. Tried to allocate 58.22 GiB (GPU 0; 22.20 GiB total capacity; 6.67 GiB already allocated; 12.59 GiB free; 8.36 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
Epoch 0:  50%|█████     | 1/2 [01:18<01:18, 78.38s/it, loss=863, split_idx=33, v_num=, train_step_loss (l1)=356.0]

```

### Environment

```
- Python: 3.9
- PyTorch Lightning: 1.7.7
- Pytorch: 1.12.1+cu116

Environment: AWS EC2 g5.16xlarge
- OS: Amazon Linux 2
- CUDA: 11.6
- GPU: NVIDIA A10G Tensor Core GPU

```

### More info

_No response_

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Closes #15057
Fixes https://github.com/Lightning-AI/lightning/issues/8732
Fixes #6539

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda

### CODE REVIEW

**Review by carmocca — COMMENTED** [review]
Looks good so far. Pushed a commit with a few fixes. Also added links in the docstring

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/utilities/migration/utils.py
Comment (carmocca) [inline src/pytorch_lightning/utilities/migration/utils.py:152]: Reuse variable and shortcut if already False

```suggestion
    target_version = Version(target)
    is_lte_max_version = max_version is None or target_version <= Version(max_version)
    return is_lte_max_version and Version(_get_version(checkpoint)) < target_version
```

**Review by carmocca — COMMENTED** [review]
LGTM. But let's wait until the fabric rename lands before merging this

**Comment by awaelchli** [comment]
How on earth am I going to rebase this.

File: src/pytorch_lightning/core/module.py
Comment (tchaton) [inline src/pytorch_lightning/core/module.py:1733]: Should we move this as a utility or get rid of it completely ?

File: src/pytorch_lightning/utilities/migration/migration.py
Comment (tchaton) [inline src/pytorch_lightning/utilities/migration/migration.py:231]: Neat !

**Review by tchaton — APPROVED** [review]
LGTM !

File: src/pytorch_lightning/loops/optimization/manual_loop.py
Comment (carmocca) [inline src/pytorch_lightning/loops/optimization/manual_loop.py:45]: Maybe this should still create a copy to maintain the old behaviour

```suggestion
            extra = training_step_output.copy()
```

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/core/module.py
Comment (carmocca) [reply src/pytorch_lightning/core/module.py:1733]: I agree with Adrian and would add a small deprecation just for this on the 1.9 branch.

The graveyard is not removed yet. There's an issue blocking the PR that I haven't been able to resolve

**Comment by Borda** [comment]
@awaelchli this shall be fixed with #16302

**Comment by awaelchli** [comment]
@Borda Nice! Thank you for jumping onto it

**Pull request merged**
